### PR TITLE
#298 | Available widgets with labels are send from backend + excluding options

### PR DESCRIPTION
**Pull request by szymon-owczarzak**
## Description

**Cogboard Mobile** is in development it is using the same backend as normal **Cogboard (Web)** that is why we need to move widget type labels from React app to the backend. Both fronends will be able to use the same set of available widget types.  

Additionally, I've implemented an option to exclude widgets that we don't need. It can be done in `/mnt/app-config.json`

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the [code style](https://github.com/Cognifide/cogboard/blob/master/CONTRIBUTING.md#coding-conventions) of this project.
- [x] My change requires a change to the documentation.
- [x] JUinit tests added
- [x] I have updated the documentation accordingly - https://github.com/Cognifide/cogboard/wiki#excluding-widgets

---

I hereby agree to the terms of the Cogboard Contributor License Agreement.

### CODE REVIEW

File: cogboard-app/src/main/kotlin/com/cognifide/cogboard/config/ConfigType.kt [hunk 1]
```diff
@@ -6,15 +6,22 @@ import com.cognifide.cogboard.config.validation.credentials.CredentialsValidator
...
     ADMIN("admin.json", AdminValidator),
-    VERSION(".version", VersionValidator);
+    VERSION(".version", VersionValidator),
+    APP_CONFIG("app-config.json", object : Validator {
+        override fun validate(config: String) = true
+    });
+
+    fun configFilePath() =
+            if (File("/data").exists()) "/data/$fileName"
+            else "${System.getProperty("user.dir")}/../mnt/$fileName"
```
Comment (szymon-owczarzak) [inline cogboard-app/src/main/kotlin/com/cognifide/cogboard/config/ConfigType.kt:24]: This is used when you run JUnit tests

File: cogboard-app/src/main/kotlin/com/cognifide/cogboard/storage/VolumeStorageFactory.kt [hunk 2]
```diff
@@ -14,6 +10,7 @@ object VolumeStorageFactory {
     fun credentials(configFile: String = CREDENTIALS.configFilePath()) = create(CREDENTIALS, configFile)
     fun boards(configFile: String = BOARDS.configFilePath()) = create(BOARDS, configFile)
     fun version(configFile: String = VERSION.configFilePath()) = create(VERSION, configFile)
+    fun appConfig(configFile: String = APP_CONFIG.configFilePath()) = create(APP_CONFIG, configFile)
```
Comment (szymon-owczarzak) [inline cogboard-app/src/main/kotlin/com/cognifide/cogboard/storage/VolumeStorageFactory.kt:13]: New storage added for excluded widgets

File: cogboard-app/src/main/kotlin/com/cognifide/cogboard/widget/WidgetIndex.kt [hunk 3]
```diff
@@ -1,53 +1,71 @@
...
+        fun create(config: JsonObject, vertx: Vertx): BaseWidget {
+            val typeName = config.getString(CC.PROP_WIDGET_TYPE)
+            val widgetType = AVAILABLE_WIDGETS
+                    .values
+                    .stream()
+                    .filter { it.simpleName == typeName }
+                    .findFirst()
+                    .orElse(WhiteSpaceWidget::class.java)
+
+            return widgetType.constructors[0].newInstance(vertx, config, BoardsConfigService()) as BaseWidget
```
Comment (szymon-owczarzak) [inline cogboard-app/src/main/kotlin/com/cognifide/cogboard/widget/WidgetIndex.kt:68]: Widgets are created using the reflection mechanism.

File: cogboard-app/src/main/resources/initData/app-config.json [hunk 4]
```diff
@@ -0,0 +1,3 @@
+{
+  "excludedWidgets": ["Example"]
```
Comment (szymon-owczarzak) [inline cogboard-app/src/main/resources/initData/app-config.json:2]: Here administrators can set which widgets are not shown in dropdown.

File: cogboard-webapp/src/components/widgets/dialogFields/WidgetTypeField.js [hunk 5]
```diff
@@ -0,0 +1,27 @@
+import React from 'react';
+import { useSelector } from 'react-redux';
+
+import { MenuItem } from '@material-ui/core';
+
+import { StyledFormControlForDropdown } from '../../styled';
+import DropdownField from '../../DropdownField';
+
+const WidgetTypeField = props => {
+  const types = useSelector(({ widgets }) => widgets.widgetTypes);
```
Comment (szymon-owczarzak) [inline cogboard-webapp/src/components/widgets/dialogFields/WidgetTypeField.js:10]: Available types come from backend now.

File: cogboard-webapp/src/components/widgets/index.js [hunk 6]
```diff
@@ -18,12 +18,10 @@ import ToDoListWidget from './types/ToDoListWidget';
 
 const widgetTypes = {
   WhiteSpaceWidget: {
-    name: 'White Space',
```
Comment (szymon-owczarzak) [inline cogboard-webapp/src/components/widgets/index.js:21]: Now names come from backend

File: cogboard-app/src/main/kotlin/com/cognifide/cogboard/widget/type/JenkinsJobWidget.kt [hunk 7]
```diff
@@ -11,8 +11,8 @@ import com.cognifide.cogboard.CogboardConstants as CC
 class JenkinsJobWidget(
     vertx: Vertx,
     config: JsonObject,
-    boardService: BoardsConfigService = BoardsConfigService()
-) : AsyncWidget(vertx, config, boardService) {
+    serv: BoardsConfigService
+) : AsyncWidget(vertx, config, serv) {
 
     private val path: String = config.getString("path", "")
```
Comment (yahorm) [inline cogboard-app/src/main/kotlin/com/cognifide/cogboard/widget/type/JenkinsJobWidget.kt:17]: could be const from cogboard const `PROP_PATH`

**Review by yahorm — DISMISSED** [review]

**Pull request merged**